Raspberry Pi Pico Clock Control Driver

andrei-edward-popa opened the pull request:
I have added driver support for clock control for Raspberry Pi Pico. Currently external clock sources are not supported by this driver.

andrei-edward-popa committed:
modules: hal_rpi_pico: Added clock sources and includes

Enabled clock sources and includes

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
dts: bindings: clock: Added clock bindings for Raspberry Pi Pico

Added clock device tree bindings for Raspberry Pi Pico platform

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
include: clock: Added clock control header files for Raspberry Pi Pico

Added clock control header files for Raspberry Pi Pico

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
boards: arm: rpi_pico: Added clock support in documentation

Added clock in supported hardware features

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
drivers: watchdog: Changed how to get xtal frequency for Raspberry Pi Pico

Changed how to get xtal frequency for Raspberry Pi Pico

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
boards: arm: rpi_pico: Enabled syscon node in device tree

Enabled syscon node in device tree
Added clock control in defconfig

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
CODEOWNERS: Added myself to codeowners for Raspberry Pi Pico clock driver

Added myself to codeowners for Raspberry Pi Pico clock driver

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
boards: arm: adafruit_kb2040: Added clock control as default in defconfig

Added clock control as default in defconfig

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
boards: arm: adafruit_kb2040: Removed xtal clock node from device tree

Removed xtal clock node from device tree

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

soburi commented on dts/bindings/clock/raspberrypi,pico-syscon.yaml:
```diff
@@ -0,0 +1,48 @@
+
+description: Raspberry Pi Pico clock controller node
+
+compatible: "raspberrypi,pico-syscon"
```
Is syscon the preferred postfix?
`syscon.yaml` exists, but it is not using that definition.
IMO, `clock-controller` (or `clock` related) postfix seems better here.

andrei-edward-popa replied:
Yeah, I think you are right, I used clocks now

MaureenHelm commented:
@andrei-edward-popa @yonsch ping

this also needs a rebase

yonsch commented on dts/bindings/clock/raspberrypi,pico-syscon.yaml:
Why is the binding name `syscon`? It would be better to just call it clock/clk/clocks, like it's called in the datasheet

andrei-edward-popa replied:
Yes, you are right, I used clocks now

yonsch commented on drivers/clock_control/clock_control_rpi_pico.c:
```diff
@@ -0,0 +1,258 @@
+
+#define DT_DRV_COMPAT raspberrypi_pico_syscon
+
+#include <zephyr/drivers/clock_control/rpi_pico_clock_control.h>
```
This is the only place this header is used (as far as I can see), can it be removed and its contents be copied here?

andrei-edward-popa replied:
A user can use the structure rpi_pico_clk_setup to setup a clock inside its application

yonsch commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -68,6 +57,34 @@
+			usb-frequency = <48000000>;
+			adc-frequency = <48000000>;
+			rtc-frequency = <46875>;
+			xtal_clk: xtal-clk {
```
Are these still needed?

andrei-edward-popa replied:
Yes, this are used inside the driver code

yonsch replied:
Okay, I see you're using them as `source_rate`. But in any case, these should be defined at the board level, as they depend on the value of the oscillator on the board (if it has one). What about choosing external clock sources and setting up the PLLs?

yonsch commented on drivers/watchdog/wdt_rpi_pico.c:
```diff
@@ -154,18 +154,19 @@ static const struct wdt_driver_api wdt_rpi_pico_driver_api = {
+#define WDT_RPI_PICO_WDT_DEVICE(idx)                                                             \
+	static const struct wdt_rpi_pico_config wdt_##idx##_config = {                           \
+		.xtal_frequency =                                                                \
+			DT_PROP(DT_CHILD(DT_NODELABEL(syscon), xtal_clk), clock_frequency)       \
```
Shouldn't that be `clk_ref`?

andrei-edward-popa replied:
Yes, I read the docs and it seems to be clk_ref

yonsch commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -59,13 +60,51 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
+
+	baudrate_div = (8 * pclk / input_baudrate);
+	baudrate_int = baudrate_div >> 7;
+	baudrate_frac = baudrate_int == 0 || baudrate_int >= 65535 ? 0 :
```
Add parentheses for clarity (also for the next statement)

andrei-edward-popa replied:
Done

yonsch commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -59,13 +60,51 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
+	uart_hw->ibrd = baudrate_int;
+	uart_hw->fbrd = baudrate_frac;
+
+	uart_hw->lcr_h |= 0;
```
Why is this needed?

andrei-edward-popa replied:
PL011 needs a dummy line control register write to latch in the divisors

yonsch replied:
So please add a comment explaining it

yonsch commented on drivers/serial/uart_rpi_pico.c:
Now that we don't use picosdk for the serial driver, you can remove [this line](https://github.com/zephyrproject-rtos/zephyr/blob/44a8b70cccafe312a1a091de48a5ba204dadd6bb/drivers/serial/Kconfig.rpi_pico#L9)

andrei-edward-popa replied:
We cannot do this because some macros are from the Pico SDK

yonsch reviewed (changes requested):
Sorry for the delay, added a few comments. @soburi @nordic-krch can you take a look?

andrei-edward-popa committed:
dts: arm: rpi_pico: Added syscon as clock controller

Added syscon as clock controller
Changed clocks for all peripherals

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
drivers: clock_control: Added clock driver for Raspberry Pi Pico

Added clock driver for Raspberry Pi Pico platform

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
drivers: spi: Changed how to get clock frequency for PL022

Changed how to get clock frequency for PL022

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
drivers: serial: Removed all function calls from Raspberry Pi Pico SDK

Removed all function calls from Raspberry Pi Pico SDK
Added functions for setting uart baudrate and format

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
boards: arm: sparkfun_pro_micro_rp2040: Added clock control as default

Added clock control as default in defconfig

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
boards: arm: sparkfun_pro_micro_rp2040: Removed xtal clock node from dts

Removed xtal clock node from device tree

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

yonsch commented on drivers/clock_control/CMakeLists.txt:
```diff
@@ -21,7 +21,7 @@ zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_NRF                 clock_cont
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_NRF_DRIVER_CALIBRATION nrf_clock_calibration.c)
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_RV32M1_PCC          clock_control_rv32m1_pcc.c)
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_INFINEON_CAT1       clock_control_ifx_cat1.c)
-
```
Something happened here with the blank lines...

yonsch commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -59,13 +60,51 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
+
+	baudrate_div = (8 * pclk / input_baudrate);
+	baudrate_int = baudrate_div >> 7;
+	baudrate_frac = (baudrate_int == 0) || (baudrate_int >= 65535) ? 0 :
```
Add even more parentheses

yonsch commented on drivers/clock_control/clock_control_rpi_pico.c:
```diff
@@ -0,0 +1,258 @@
+		[RPI_PICO_CLOCK_PERI] = {
+			.source = 0,
+			.aux_source = CLOCKS_CLK_PERI_CTRL_AUXSRC_VALUE_CLK_SYS,
+			.source_rate = DT_INST_PROP(0, sys_frequency),
```
Is this correct?

yonsch commented on drivers/clock_control/clock_control_rpi_pico.c:
```diff
@@ -0,0 +1,258 @@
+#include <hardware/clocks.h>
+
+#define CTRL_ENABLE_BITS CLOCKS_CLK_GPOUT0_CTRL_ENABLE_BITS
+#define FIXED_CLOCK_FREQ(node_id, child) DT_PROP(DT_CHILD(node_id, child), clock_frequency)
```
This seems to be written specifically to access the children of `clocks`, so maybe it should be rewritten to something like:
```
#define FIXED_CLOCK_FREQ(clock) DT_PROP(DT_CHILD(DT_NODELABEL(clocks), clock), clock_frequency)
```

yonsch reviewed (changes requested):
A few questions, most importantly: how will clock sources be supported?
